### ISSUE TITLE
Inconsistent behavior between `_io.text_encoding` and `_pyio.text_encoding`

### ISSUE THREAD

**Issue by donbarbos** [python/cpython#127611]
# Bug report

### Bug description:

```python
>>> import _io, _pyio
>>> _pyio.text_encoding(None, stacklevel=2)
'locale'
>>> _io.text_encoding(None, stacklevel=2)
Traceback (most recent call last):
  File "<python-input-54>", line 1, in <module>
    _io.text_encoding(None, stacklevel=2)
    ~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^
TypeError: _io.text_encoding() takes no keyword arguments
```

### CPython versions tested on:

CPython main branch

**Comment by tomasr8** [comment]
Indeed, the C function signature is marked as positional-only:

https://github.com/python/cpython/blob/e51da64ac3bc6cd45339864db32d05115af39ead/Modules/_io/_iomodule.c#L457-L460

This was originally added when [PEP 597](https://peps.python.org/pep-0597/) was implemented:
https://github.com/python/cpython/pull/19481/files#diff-0501b143a2f566e8a760abc20a88fd2533b1f230e5322831f400a0263b8c0253R511-R514

However, I didn't find any comments in the PR explaining why this is so. In addition, the [example implementation in the PEP](https://peps.python.org/pep-0597/#io-text-encoding) also allows keyword arguments.

cc @methane as the original author

**Comment by methane** [comment]
I just didn't care it well.

The doc uses positional only argument. It is official specification than the PEP.
https://docs.python.org/3.13/library/io.html#io.text_encoding

Unless clear use case for keyword argument here, how about just prohibit keyword argument on `_pyio` too?

**Comment by donBarbos** [comment]
Firstly, the documentation doesn't say anything about how to use the `stacklevel` argument. Yes, the positional argument is used in the example, but only the `encoding` argument is passed there.
Also, the C signature of the `_io.open` function is marked exactly the same like `stacklevel` argument (but these are keyword arguments).
https://github.com/python/cpython/blob/e51da64ac3bc6cd45339864db32d05115af39ead/Modules/_io/_iomodule.c#L74-L82

Secondly, if we remove `stacklevel` argument as keyword argumnet in `_pyio` implementation and leave it as positional argument, it will break backward compatibility.

I just suggest to consider `stacklevel` argument as a keyword argument.

cc @methane, @tomasr8

**Comment by methane** [comment]
> Firstly, the documentation doesn't say anything about how to use the `stacklevel` argument. Yes, the positional argument is used in the example, but only the `encoding` argument is passed there.

<img width="470" alt="image" src="https://github.com/user-attachments/assets/f9d9adb1-4e4f-4644-840e-856b89ad00f3">

The `/` means all arguments are positional-only.

**Comment by methane** [comment]
> Secondly, if we remove `stacklevel` argument as keyword argumnet in `_pyio` implementation and leave it as positional argument, it will break backward compatibility.

`_pyio` is very minor module. Almost no one use it except testing purpose.

* Removing keyword support from _pyio needs only changelog.
* Adding keyword support to _io needs document update (e.g. `versionchanged`) and what's new etry.

This is why I am proposing just remove it unless there is clear benefits.

**Comment by donBarbos** [comment]
@methane, sounds good enough to me. I have no more objections to fixing the python implementation.
I'm ready to send PR.

**Comment by tomasr8** [comment]
Requiring positional-only arguments in `pyio` is definitely easier but I lean more towards allowing kw arguments in the C implementation.

For one, it's not a breaking change, unlike changing `pyio` even if very few people are using it. There's also no good reason for enforcing positional-only arguments and I'd even say it promotes bad habits because it forces you to write code like `text_encoding(encoding, 2)` with a magic constant, rather than `text_encoding(encoding, stacklevel=2)`.

**Comment by methane** [comment]
FWIW, this function is not for regular users.
Good habit is do not use this function at all. Use `encoding="utf-8"` as default encoding instead.

This function is for almost only stdlib. 3rd party libraries may need it too, but it is rare.
Additionally, using stacklevel other than 2 is more rare. It is really rare.

That is why I am reluctant to make API improvements that would require a what's new entry and `versionchanged` note.
But If you still wish to make improvements, I am not opposed.

**Comment by methane** [comment]
No. PyPy has own builtin `_io` module.

**Comment by donBarbos** [comment]
I think it's worth mentioning that `encoding` argument also cannot be passed as keyword.

And I agree that the practice of using keyword arguments may be a better (readable) option.

So what will the consensus be?

**Comment by methane** [comment]
No consensus yet. I don't want to change this function. This function is almost only for stdlib. No need to improve usability for many users.

**Comment by picnixz** [comment]
We have a similar case for `functools.reduce` https://github.com/python/cpython/issues/121676 so whatever decision we make here should be made out there for consistency IMO (and vice-versa). They both suffer the same issue, namely the Python implementation accepts keyword arguments while the C implementation does not and the docs say "positional-only".

**Comment by methane** [comment]
They are very different functions. It seems a foolish consistency.

**Comment by picnixz** [comment]
They may be different but IMO the issue is the same. A publicly documented function that behaves differently with the C and the Python implementation. But if you don't want the parallel, I'm ok (and would therefore rather not change `text_encoding` at all).

**Comment by tomasr8** [comment]
I agree with @picnixz. If we don't want to change the C implementation, I'd also prefer to not do anything since there seems to be very little benefit to restricting the Python implementation for the sake of consistency.

**Comment by donBarbos** [comment]
I found another issue #89979 on a related topic, maybe it's worth moving this whole discussion here (but it's not progressing there)

This problem is common but no one takes the initiative to solve it.
Maybe we need to ask someone to give us the go-ahead to make more consistency in both implementations :-)

**Comment by picnixz** [comment]
In general, the Python implementation is aligned with the C implementation when the Python implementation is used by other implementations (e.g., functools for pypy). If it's not, we tend not to bother too much 1) if existing code that works with the C implementation works, or 2) if it is half public API or can be replaced by a more modern approach. So, indeed, the functools issue is different (because there is a downstream user using the Python implementation as its sole alternative); sorry for that, this was my bad.

Note that in both cases, we document the API with positional-only, so whatever implementation we choose, it's actually ok if users follow the docs. Now, in the datetime case, the documentation does not put restrictions so people could assume that keyword arguments are allowed while they are not in some implementations (so following the docs could break their code on some Python implementations). It's again a slightly different issue compared to this one and the functools one.

>  Maybe we need to ask someone to give us the go-ahead to make more consistency in both implementations :-)

Inada is a core dev, and likely the code owner of unicode-related stuff. So that "someone" would be them. Alex (another core dev, who also opened the datetime issue) also reacted to https://github.com/python/cpython/issues/127611#issuecomment-2522699292 with a 👍, which I understand as "let's not do it".

**Comment by erlend-aasland** [comment]
Closing, as Inada-san clearly prefer status quo.
